pkg/model/iam/pod_roles.go

```diff
@@ -19,10 +19,50 @@ package iam
 import (
 	"fmt"
 
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/kops/pkg/apis/kops"
 	"k8s.io/kops/pkg/featureflag"
+	"k8s.io/kops/pkg/wellknownusers"
 )
 
+// ServiceAccountRole holds a pod-specific IAM policy
+type ServiceAccountRole string
+
+const (
+	// ServiceAccountRoleEmpty indicates we are not using a ServiceAccountRole
+	ServiceAccountRoleEmpty ServiceAccountRole = ""
+	// ServiceAccountRoleDNSController is the role used by the dns-controller pods
+	ServiceAccountRoleDNSController ServiceAccountRole = "dns-controller"
+)
+
+// PodOrNodeRole is used to specify generation of permissions for either a pod or a node
+type PodOrNodeRole struct {
```

Comment:
I gave it a go, the problem was that because go interfaces aren't explicit, I found the end result to be really unintuitive and unobvious.  You end up with these weird interfaces that are basically `void *`.  I suggest we try it again in go2.

Here's the core interface I tried - maybe you had a different one in mind?

```
// IAMRole is implemented by our IAM scoping objects: ServiceAccountRole and InstanceGroupRole.
type IAMRole interface {
	// HasServiceAccountRole returns true if we include the specified ServiceAccountRole.
	HasServiceAccountRole(role ServiceAccountRole) bool

	// HasNodeRole returns true if we include the specified NodeRole.
	HasNodeRole(role kops.InstanceGroupRole) bool
}
```

Reply:
I was thinking of pushing the relevant `buildAWSIAMRolePolicy()` logic under an interface method. And the `HasServiceAccountRole` method would return the name of any service account.

Basically I'm applying the principle that code that does type assertions (or tests for existence of union fields) suggests that the branches under the test might be better as an interface method.